File: sample/src/main/java/io/erva/sample/base/BaseSampleActivity.java
package io.erva.sample.base;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.List;

import io.erva.celladapter.CellAdapter;
import io.erva.sample.DividerItemDecoration;
import io.erva.sample.R;
import io.erva.sample.base.cell.AlphaCell;
import io.erva.sample.base.cell.BetaCell;
import io.erva.sample.base.cell.GammaCell;
import io.erva.sample.base.model.AlphaModel;
import io.erva.sample.base.model.BetaModel;
import io.erva.sample.base.model.GammaModel;

public class BaseSampleActivity extends AppCompatActivity {

    private RecyclerView recyclerView;
    private CellAdapter adapter;
    private Toast toast;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_with_recycler_view);

        recyclerView = (RecyclerView) findViewById(R.id.recycler_view);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        recyclerView.addItemDecoration(new DividerItemDecoration(this));

        adapter = new CellAdapter();
        adapter.registerCell(AlphaModel.class, AlphaCell.class, new AlphaCell.Listener() {

            @Override
            public void onPressOne(AlphaModel model) {
                showToast(String.format("%s%npress button %d", model.getAlpha(), 1));
            }

            @Override
            public void onPressTwo(AlphaModel model) {
                showToast(String.format("%s%npress button %d", model.getAlpha(), 2));
            }

            @Override
            public void onCellClicked(AlphaModel model) {
                showToast(model.getAlpha());
            }
        });
        adapter.registerCell(BetaModel.class, BetaCell.class, new BetaCell.Listener() {
            @Override
            public void onCellClicked(BetaModel model) {
                showToast(model.getBeta());
            }
        });
        adapter.registerCell(GammaModel.class, GammaCell.class, null);
        recyclerView.setAdapter(adapter);

        List items = new ArrayList();
        for (int i = 0; i <= 33; i++) {
            items.add(new AlphaModel(String.format("AlphaModel %d", i)));
            items.add(new BetaModel(String.format("BetaModel %d", i)));
            items.add(new GammaModel(String.format("GammaModel %d", i)));
        }

        adapter.setItems(items);
        adapter.notifyDataSetChanged();
    }

    @Override
    protected void onStop() {
        super.onStop();
        dismissToast();
    }

    private void showToast(String text) {
        dismissToast();
        toast = Toast.makeText(this, text, Toast.LENGTH_SHORT);
        toast.show();
    }

    private void dismissToast() {
        if (toast != null) toast.cancel();
    }
}

File: sample-kotlin/src/main/kotlin/io/erva/sample/base/BaseSampleActivity.kt
package io.erva.sample.base

import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.LinearLayoutManager
import android.widget.Toast
import io.erva.celladapter.CellAdapter
import io.erva.sample.DividerItemDecoration
import io.erva.sample.R
import io.erva.sample.base.cell.AlphaCell
import io.erva.sample.base.cell.BetaCell
import io.erva.sample.base.cell.GammaCell
import io.erva.sample.base.model.AlphaModel
import io.erva.sample.base.model.BetaModel
import io.erva.sample.base.model.GammaModel
import kotlinx.android.synthetic.main.activity_with_recycler_view.*

class BaseSampleActivity : AppCompatActivity() {

    private var toast: Toast? = null
    private var adapter: CellAdapter = CellAdapter().let {
        it.cell(AlphaCell::class) {
            item(AlphaModel::class)
            listener(object : AlphaCell.Listener {

                override fun onPressOne(item: AlphaModel) {
                    showToast(String.format("%s%npress button %d", item.alpha, 1))
                }

                override fun onPressTwo(item: AlphaModel) {
                    showToast(String.format("%s%npress button %d", item.alpha, 2))
                }

                override fun onCellClicked(item: AlphaModel) {
                    showToast(item.alpha)
                }
            })
        }
        it.cell(BetaCell::class) {
            item(BetaModel::class)
            listener(object : BetaCell.Listener {
                override fun onCellClicked(item: BetaModel) {
                    showToast(item.beta)
                }
            })
        }
        it.cell(GammaCell::class) {
            item(GammaModel::class)
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_with_recycler_view)

        recycler_view.layoutManager = LinearLayoutManager(this)
        recycler_view.addItemDecoration(DividerItemDecoration(this))
        recycler_view.adapter = adapter

        for (i in 0..33) {
            adapter.items.add(AlphaModel(String.format("AlphaModel %d", i)))
            adapter.items.add(BetaModel(String.format("BetaModel %d", i)))
            adapter.items.add(GammaModel(String.format("GammaModel %d", i)))
        }
        adapter.notifyDataSetChanged()
    }

    override fun onStop() {
        super.onStop()
        dismissToast()
    }

    private fun showToast(text: String) {
        dismissToast()
        toast = Toast.makeText(this, text, Toast.LENGTH_SHORT)
        toast?.show()
    }

    private fun dismissToast() {
        toast?.cancel()
    }
}